File: go/internal/tools/tools.go
// Tool definitions + dispatch via the shared rust binary `teleia-tools-bin`.
package tools

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"os/exec"
	"strings"

	"github.com/foolish-dev/Teleia/go/internal/llm"
)

const binary = "teleia-tools-bin"

func Definitions() []llm.ToolDef {
	out, err := exec.Command(binary, "defs").Output()
	if err != nil {
		panic(fmt.Sprintf("%s defs: %v", binary, err))
	}
	var defs []llm.ToolDef
	if err := json.Unmarshal(out, &defs); err != nil {
		panic(fmt.Sprintf("parse %s defs: %v", binary, err))
	}
	return defs
}

// Dispatch runs a tool via teleia-tools-bin. The binary encodes tool failures
// in stdout (always exits 0), so wrappers don't need to distinguish dispatch
// errors from non-zero bash exit codes — those are already in the output.
// A non-zero exit here means the binary itself failed (e.g., not on PATH).
func Dispatch(ctx context.Context, name, arguments string) (string, error) {
	cmd := exec.CommandContext(ctx, binary, "run", name)
	cmd.Stdin = strings.NewReader(arguments)
	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr
	if err := cmd.Run(); err != nil {
		return fmt.Sprintf("error: %s", strings.TrimSpace(stderr.String())), nil
	}
	return stdout.String(), nil
}

File: bun/src/tools.ts
// Tool definitions + dispatch via the shared rust binary `teleia-tools-bin`.
import type { ToolDef } from "./llm"

const BINARY = "teleia-tools-bin"

export function definitions(): ToolDef[] {
  const result = Bun.spawnSync([BINARY, "defs"], { stdout: "pipe", stderr: "pipe" })
  if (result.exitCode !== 0) {
    throw new Error(`${BINARY} defs: ${result.stderr.toString().trim()}`)
  }
  return JSON.parse(result.stdout.toString()) as ToolDef[]
}

// Run a tool via teleia-tools-bin. The binary always exits 0 and encodes
// tool failures as "error: ..." in stdout. A non-zero exit here means the
// binary itself failed (e.g., not on PATH).
export function dispatch(name: string, argumentsJson: string): string {
  const result = Bun.spawnSync([BINARY, "run", name], {
    stdin: new TextEncoder().encode(argumentsJson),
    stdout: "pipe",
    stderr: "pipe",
  })
  if (result.exitCode !== 0) {
    return `error: ${result.stderr.toString().trim()}`
  }
  return result.stdout.toString()
}
